src/fsharp/FSharp.Core/map.fs

```diff
@@ -619,6 +702,45 @@ namespace Microsoft.FSharp.Collections
                 | _ -> 
                     invalidArg "obj" (SR.GetString(SR.notComparable))
 
+        interface System.IComparable<Map<'Key,'Value>> with
+            member m.CompareTo other =
+                // REVIEW: Docs for IComparable say implementations shouldn't raise exceptions for null arguments.
+                (m, other)
+                ||> Seq.compareWith (fun (kvp1 : KeyValuePair<_,_>) (kvp2 : KeyValuePair<_,_>) ->
+                    let c = comparer.Compare(kvp1.Key,kvp2.Key)
+                    if c <> 0 then c else Unchecked.compare kvp1.Value kvp2.Value)
+
+        interface System.IEquatable<Map<'Key,'Value>> with
+            member this.Equals other =
+                // Not equal if the other instance is null.
+                not (Object.ReferenceEquals (null, other)) && this.EqualsImpl other
+
+#if FX_ATLEAST_45
+        interface IReadOnlyCollection<KeyValuePair<'Key, 'Value>> with
+            member s.Count = s.Count
+
+        interface IReadOnlyDictionary<'Key, 'Value> with
+            member m.Item
+                with get x = m.[x]
+
+            // REVIEW: this implementation could avoid copying the Keys to an array
+            member s.Keys = ([| for kvp in s -> kvp.Key |] :> IEnumerable<'Key>)
```

Comment:
Can we somehow rely on the same member .Keys and .Values which are implemented above instead of duplicating the code?

maybe make an inline function to be used in both places instead?